### ISSUE TITLE
Any signal can unfreeze a frozen container

### ISSUE THREAD

**Issue by devimc** [moby/moby#37137]
**Steps to reproduce the issue:**
*Terminal A*
`docker run --rm --name foo --runtime=runc -ti debian bash`

*Terminal B*
`docker pause foo`

*Terminal A*
Type anything, container must be frozen 

*Terminal B*
`docker kill foo -s SIGUSR1`

*Terminal A*
Container has been thawed

**Describe the results you received:**
Frozen containers can be thawed with a signal, I thought only `docker resume` could do this

**Describe the results you expected:**
a signal shouldn't unfreeze a container, right?

**Output of `docker version`:**

```
Client:
 Version:      18.03.1-ce
 API version:  1.37
 Go version:   go1.9.5
 Git commit:   9ee9f40
 Built:        Thu Apr 26 07:21:07 2018
 OS/Arch:      linux/amd64
 Experimental: false
 Orchestrator: swarm

Server:
 Engine:
  Version:      18.03.1-ce
  API version:  1.37 (minimum version 1.12)
  Go version:   go1.9.5
  Git commit:   9ee9f40
  Built:        Thu Apr 26 07:24:55 2018
  OS/Arch:      linux/amd64
  Experimental: false
```

**Output of `docker info`:**

```
Containers: 2
 Running: 1
 Paused: 0
 Stopped: 1
Images: 56
Server Version: 18.03.1-ce
Storage Driver: overlay2
 Backing Filesystem: extfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: cc-runtime kata-runtime runc
Default Runtime: kata-runtime
Init Binary: docker-init
containerd version: 773c489c9c1b21a6d78b5c538cd395416ec50f88
runc version: 1252058fc0833ece7fe407672543179bf9adc21e-dirty (expected: 4fc53a81fb7c994640722ac585fa9ca548971871)
init version: 949e6fa
Security Options:
 seccomp
  Profile: default
Kernel Version: 4.16.7-100.fc26.x86_64
Operating System: Fedora 26 (Twenty Six)
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 15.54GiB
Name: XXXXXXXXXXXXXXX
ID: XXXXXXXXXXXXXXXXXXx
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 29
 Goroutines: 52
 System Time: 2018-05-23T14:02:30.464879002-05:00
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
```

**Comment by thaJeztah** [comment]
The container is unpaused to send the signal;  https://github.com/moby/moby/blob/3a633a712c8bbb863fe7e57ec132dd87a9c4eff7/daemon/kill.go#L73-L86

This was added in https://github.com/moby/moby/commit/c3feb046b9b1eb606cc7e853d020799eb3800e0e (https://github.com/moby/moby/pull/34027) with the intent to allow killing a paused container (looks like we may need some documentation updates for that)

@cpuguy83 the intent was to allow killing the container; I wonder if allowing other signals was the intent as well, or a side effect of that change?

**Comment by cpuguy83** [comment]
It should only unpause when the configured stop signal is sent.

**Comment by thaJeztah** [comment]
Ah, right, this line should take care of that https://github.com/moby/moby/blob/3a633a712c8bbb863fe7e57ec132dd87a9c4eff7/daemon/kill.go#L79 🤔

**Comment by devimc** [comment]
I think right behaviour is to don't kill a paused container, same as `freezer` cgroup

**Comment by thaJeztah** [comment]
Problem could be if the container doesn't have a stopsignal defined, https://github.com/moby/moby/blob/3a633a712c8bbb863fe7e57ec132dd87a9c4eff7/daemon/kill.go#L74

in that case we get in the "else", and any signal will unpause (if I see it correctly)

https://github.com/moby/moby/blob/3a633a712c8bbb863fe7e57ec132dd87a9c4eff7/daemon/kill.go#L83-L85

**Comment by devimc** [comment]
Probably the issue is not in docker, see https://github.com/opencontainers/runc/blob/3a079311a7e9afeab4b722616cefd2f9b4129104/libcontainer/init_linux.go#L475-L496

**Comment by devimc** [comment]
@thaJeztah @cpuguy83  frozen processes can't receive signals (cgroups), even KILL signal is ignored, why docker containers can? I think `docker kill` should check first if the container is running (not paused) before sending any signal, and `docker rm` should resume the container (if paused) before killing the container. what do you think?

**Comment by thaJeztah** [comment]
Correct; that's what the code does that I linked to; it sends a signal and un-pauses the process

**Comment by thaJeztah** [comment]
Well; that part is a couple of lines down in the same function

**Comment by devimc** [comment]
ook, then something is broken :smile:

**Comment by thaJeztah** [comment]
My suspicion is it's in that if/else condition, but haven't had time yet to try and work on a patch

**Comment by arpitjindal97** [comment]
I changed the code this

https://github.com/moby/moby/blob/9c2c887b125df56e622d20ebbbe626c55b61e44a/daemon/kill.go#L83-L86


to

```
} else if syscall.Signal(sig) == syscall.SIGKILL {
    container.ExitOnNext()
    unpause = container.Paused
} else {
    return nil
}
```

- If the SIGSTOP signal is sent and stop signal is configured then the `if` part will be executed and container will be deleted
- It will process SIGKILL signal in `else if` and a paused container will be deleted
- It will ignore any other signal and thus container will be remain paused

**Comment by thaJeztah** [comment]
The  `return nil` may not be correct; I seem to recall that signalling a paused container is ok, but it just won't respond to it while it's paused, and will receive the signal when it's un-paused.

**Comment by devimc** [comment]
@thaJeztah actually it won't receive the signal when it's un-paused, I already tested.
Container should be un-paused before sending any signal, I think so.
